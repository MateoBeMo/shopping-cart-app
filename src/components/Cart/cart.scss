@import 'styles/styles-utils.scss';

.cart__wrapper {
    padding: 16px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height: 100%;
    .cart__product-list {
        padding: 8px;
        width: 100%;
        display: flex;
        flex-direction: column;
        align-items: center;
        margin-bottom: 24px;
        overflow: auto;
        .cart__product {
            display: flex;
            width: 300px;
            margin-bottom: 40px;
            .cart__image {
                width: 100px;
                height: 100px;
                object-fit: cover;
            }
            .cart__title {
                display: flex;
                flex-direction: column;
                justify-content: space-between;
                align-items: flex-start;
                margin: 16px 16px 0 16px;
                svg {
                    cursor: pointer;
                    user-select: none;
                }
                .cart__total {
                    margin: 0 8px;
                    user-select: none;
                }
            }
            .cart__price {
                font-weight: bold;
                align-items: center;
                display: flex;
            }
        }
    }

    .cart__checkout {
        display: flex;
        justify-content: center;
        align-items: center;
        margin-bottom: 32px;
        button {
            margin-right: 16px;
        }
        span {
            font-weight: bold;
        }
    }
}
