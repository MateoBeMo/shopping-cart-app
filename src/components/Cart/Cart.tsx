import React, { useEffect, useState } from 'react';
/**Libraries */
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
/**Components */
import Button from 'components/Button';
/**Styles */
import './cart.scss';
import { ProductModel } from 'models/product.model';

type CartProps = {
    products: ProductModel[];
    handleRemoveItem: (product: ProductModel) => void;
    handleUpdateItem: (product: ProductModel) => void;
    handleCheckout: (products: ProductModel[]) => void;
};

const Cart = ({ products, handleRemoveItem, handleCheckout, handleUpdateItem }: CartProps) => {
    const [cartProducts, setCartProducts] = useState<ProductModel[]>([]);
    const [totalPrice, setTotalPrice] = useState<number>(0);

    useEffect(() => {
        setCartProducts(products.map(obj => ({ ...obj, total: obj.total || 1 })));
    }, [products]);

    useEffect(() => {
        setTotalPrice(
            cartProducts.reduce((totalPrice, product) => {
                const totalProduct = product.price * (product.total || 1);
                totalPrice += totalProduct;
                return totalPrice;
            }, 0),
        );
    }, [cartProducts]);

    const addOneProduct = (product: ProductModel) => {
        if (product.total && product.stock > product.total) {
            product.total += 1;
            handleUpdateItem(product);
            setCartProducts(prevCartProducts => [...prevCartProducts]);
        }
    };
    const removeOneProduct = (product: ProductModel) => {
        if (product.total && product.total > 1) {
            product.total -= 1;
            handleUpdateItem(product);
            setCartProducts(prevCartProducts => [...prevCartProducts]);
        } else {
            handleRemoveItem(product);
        }
    };
    const checkout = () => {
        handleCheckout(cartProducts);
        setCartProducts([]);
    };
    return (
        <div className="cart cart__wrapper">
            <div className="cart__product-list">
                {cartProducts.map(product => {
                    const { image_url, productName, price, total } = product;
                    return (
                        <div className="cart__product" key={product.id}>
                            <img className="cart__image" src={image_url} alt={productName} />
                            <div className="cart__title">
                                <span>{productName}</span>
                                <div>
                                    <span data-testid="minus-icon" onClick={() => removeOneProduct(product)}>
                                        <FontAwesomeIcon icon="minus" />
                                    </span>
                                    <span className="cart__total"> {total} </span>
                                    <span data-testid="plus-icon" onClick={() => addOneProduct(product)}>
                                        <FontAwesomeIcon icon="plus" />
                                    </span>
                                </div>
                            </div>
                            <div className="cart__price">
                                <span>{price}$</span>
                            </div>
                        </div>
                    );
                })}
            </div>
            <div className="cart__checkout">
                <Button onClick={checkout}> CHECKOUT </Button>
                <span>{totalPrice}$</span>
            </div>
        </div>
    );
};

export default Cart;
