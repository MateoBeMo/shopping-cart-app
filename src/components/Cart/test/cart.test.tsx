import React from 'react';
/**Components */
import Cart from '../Cart';
/**Testing library */
import { render, fireEvent, cleanup } from '@testing-library/react';
import { axe } from 'jest-axe';
import { ProductModel } from 'models/product.model';

type CartProps = {
    products: ProductModel[];
    handleRemoveItem: (product: ProductModel) => void;
    handleCheckout: (products: ProductModel[]) => void;
};

const handleRemoveItemMock = jest.fn();
const handleCheckoutMock = jest.fn();

const renderCart = (props: Partial<CartProps> = {}) => {
    const defaultProps: CartProps = {
        products: [
            {
                id: "c34287bf-9cb7-40a7-ad07-6534e45a6868",
                image_url: 'https://dummyimage.com/400x400/bc27b5/000&text=Gorgeous Frozen Chair',
                stock: 28,
                productName: 'Gorgeous Frozen Chair',
                price: 47,
                productDescription:
                    'Harum modi sunt. Voluptatem ut molestiae consequatur. Ea omnis architecto laboriosam accusantium reiciendis corporis exercitationem ad dolor. Fugit autem placeat voluptas sint aut aliquam sed. Totam fuga nesciunt rerum voluptatibus. Voluptatibus voluptates vel ut et temporibus perferendis laboriosam accusamus.',
                favorite: '1',
            },
            {
                id: "2af15456-ae9e-4c44-a19d-43f852dbb94f",
                image_url: 'https://dummyimage.com/400x400/4b5148/000&text=Gorgeous Rubber Mouse',
                stock: 22,
                productName: 'Gorgeous Rubber Mouse',
                price: 46,
                productDescription:
                    'Eaque voluptas vero magni eum culpa quidem. Dolorum odit atque eaque repudiandae. Occaecati doloribus esse eius et sit iste dignissimos voluptatem. Occaecati commodi eligendi voluptatum. Et iste dolor qui beatae quis ab non. Occaecati saepe praesentium itaque qui soluta asperiores.',
                favorite: '1',
            },
        ],
        handleRemoveItem: handleRemoveItemMock,
        handleCheckout: handleCheckoutMock,
    };
    return render(<Cart {...defaultProps} {...props} />);
};

describe('<Cart>', () => {
    afterEach(() => {
        cleanup();
    });

    test('Snapshot renders => it should rendered with the provided props', () => {
        const { container } = renderCart();

        expect(container).toMatchSnapshot();
    });

    
    test('It should fire handleRemoveItem event on click over minus icon when 1 product left', () => {
        const { getAllByTestId } = renderCart()

        const minusIcons = getAllByTestId('minus-icon');

        fireEvent.click(minusIcons[0]);

        expect(handleRemoveItemMock).toHaveBeenCalled();
        expect(handleRemoveItemMock).toHaveBeenCalledTimes(1);
    });

    test('It should fire handleCheckout event on click over checkout button', async() => {
        const { findByText } = renderCart()

        const checkoutButton = await findByText(/checkout/i);

        fireEvent.click(checkoutButton);

        expect(handleCheckoutMock).toHaveBeenCalled();
        expect(handleCheckoutMock).toHaveBeenCalledTimes(1);
    });

    test('should not have basic accessibility issues', async () => {
        const { container } = renderCart();
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});
