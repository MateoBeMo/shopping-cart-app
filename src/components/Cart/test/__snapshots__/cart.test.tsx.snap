// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart> Snapshot renders => it should rendered with the provided props 1`] = `
<div>
  <div
    class="cart cart__wrapper"
  >
    <div
      class="cart__product-list"
    >
      <div
        class="cart__product"
      >
        <img
          alt="Gorgeous Frozen Chair"
          class="cart__image"
          src="https://dummyimage.com/400x400/bc27b5/000&text=Gorgeous Frozen Chair"
        />
        <div
          class="cart__title"
        >
          <span>
            Gorgeous Frozen Chair
          </span>
          <div>
            <span
              data-testid="minus-icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-minus fa-w-14 "
                data-icon="minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="cart__total"
            >
               
              1
               
            </span>
            <span
              data-testid="plus-icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus fa-w-14 "
                data-icon="plus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          class="cart__price"
        >
          <span>
            47
            $
          </span>
        </div>
      </div>
      <div
        class="cart__product"
      >
        <img
          alt="Gorgeous Rubber Mouse"
          class="cart__image"
          src="https://dummyimage.com/400x400/4b5148/000&text=Gorgeous Rubber Mouse"
        />
        <div
          class="cart__title"
        >
          <span>
            Gorgeous Rubber Mouse
          </span>
          <div>
            <span
              data-testid="minus-icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-minus fa-w-14 "
                data-icon="minus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="cart__total"
            >
               
              1
               
            </span>
            <span
              data-testid="plus-icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus fa-w-14 "
                data-icon="plus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          class="cart__price"
        >
          <span>
            46
            $
          </span>
        </div>
      </div>
    </div>
    <div
      class="cart__checkout"
    >
      <button
        class=" button"
        role="button"
      >
         CHECKOUT 
      </button>
      <span>
        93
        $
      </span>
    </div>
  </div>
</div>
`;
